### ISSUE TITLE
NPM types: TSX warnings about JSX.IntrinsicElements in React

### ISSUE THREAD

**Issue by oscarotero** [denoland/deno#16653]
I have a TSX project with the following `deno.json` configuration:

```json
{
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "npm:react"
  }
}
```

In VS Code I see the following warning:

<img width="1096" alt="imaxe" src="https://user-images.githubusercontent.com/377873/202049120-61609e84-1188-4171-9d3a-40803f3ba78d.png">

This doesn't happen if the JSX import source is preact (`npm:preact`).

I think the reason is because the preact NPM package includes the types (https://github.com/preactjs/preact/blob/master/src/index.d.ts) so Deno can load them.

React needs to fetch the types from `@types/react` and this is not implemented in Deno. In the NPM roadmap (https://github.com/denoland/deno/issues/15960) I can see the uncompleted task `automatic acquisition of @types package for type checking`. 

Is there any plans to implement this at anytime soon? I think it should be quite useful. At this moment, the only way to use React in Deno without TSX errors is by importing from esm.sh.

**Comment by bartlomieju** [comment]
It is planned, but currently no ETA. Have you tried adding `// @deno-types="npm:@types/react"` as an escape hatch?

**Comment by oscarotero** [comment]
It doesn't work:
<img width="575" alt="imaxe" src="https://user-images.githubusercontent.com/377873/202051785-d24ff39e-87c7-49c3-bc31-e794632dfdd0.png">

Not sure if I have to add this in another place.

**Comment by dsherret** [comment]
Yeah, that wouldn't work because there's no import to set that on. I believe the way to do this would be to add a triple slash types reference directive:

```
/// <reference types="npm:@types/react@18" />
```

...or do...

```
import type {} from "npm:@types/react@18";
```

But then I get the error:

> 'React' refers to a UMD global, but the current file is a module. Consider adding an import instead.

...which only occurs in the LSP... not sure why.

**Comment by dsherret** [comment]
Bartek's latest commit (300fd07fad3dee74730517ba91b2e7314632fc3f) resolves that issue I was seeing in my last comment. Adding a tiple slash directive reference or `import type` for `npm:@types/react@18` seems to be the way to go here. I did come across #16654, which I will look into tomorrow.

**Comment by Oyelowo** [comment]
@dsherret
Has this now been resolved in the latest release - 1.28.1?  
I'm asking as I am still running into the same issue - code runs but the type-checkings fail as mentioned above.

I'm also getting the below error on all my imports which causes types to fail altogether. 
One other source of confusion is the `deno-react-template` does not use `import_map.json` at all and all imports are done in `vite.config`. Which is the way to go?

<img width="824" alt="Screenshot 2022-11-17 at 18 21 10" src="https://user-images.githubusercontent.com/31687368/202500533-b729e0c5-8ab9-46e4-a11a-c813deb89379.png">

 thanks for your work on this! 👍

**Comment by Oyelowo** [comment]
I resolved temporarily this by using `import_map` with esm.sh whole `vite` usings `npm:` mapping for building and running

**Comment by sant123** [comment]
This is my workaround for server side rendering. Hope it helps:

```ts
// @deno-types="npm:@types/react"
import React from "npm:react";
// @deno-types="npm:@types/react-dom/server"
import ReactDOM from "npm:react-dom/server";

function App() {
  return <h1>Hello world!</h1>;
}

console.log(ReactDOM.renderToString(<App />));

```

**Comment by jimisaacs** [comment]
`renderToString` is still resolving as `any` with the above ☝️ for me. It doesn't seem like `// @deno-types="npm:@types/react-dom/server"` is working as expected.

**Comment by dsherret** [comment]
@jimisaacs maybe related https://github.com/denoland/deno/issues/17012 -- this was just fixed and a test was added yesterday to prevent regressions. It seems to work in canary (`deno upgrade --canary`).

**Comment by jimisaacs** [comment]
@dsherret  Nice! Good to know it seems to be addressed. Though unfortunately it's a little difficult to try because of the following `error: Canary builds are not available for M1`

I was hoping to see if all this was related, i.e. the same issue was causing the following error with `deno test`:
```
error: Uncaught Error: Error resolving package config for 'npm:@types/react-dom@18/server'.: [ERR_INVALID_PACKAGE_TARGET] Invalid "exports" target {"default":"./server.d.ts"} defined for './server' in the package config /my-deno-dir/cache/npm/registry.npmjs.org/@types/react-dom/18.0.9package.json imported from file:///my-deno-dir/cache/npm/registry.npmjs.org/@types/react-dom/18.0.9/; target must start with "./"
```

**Comment by jimisaacs** [comment]
🙇  `--no-check` did the trick, like 99.9% confident it's all the same issue. Thank you!

**Comment by sant123** [comment]
@dsherret @jimisaacs just tested with canary and looks good!!

![image](https://user-images.githubusercontent.com/7959437/208154597-c399354c-e06e-4417-aff0-434b261b265a.png)

**Comment by dylanblokhuis** [comment]
This will also work:
```tsx
import type {} from "npm:@types/react"
import type {} from "npm:@types/react-dom"
```
The LSP will correctly pickup the types

**Comment by nayeemrmn** [comment]
Ref https://github.com/denoland/deno/issues/20455
